ociclient/manifest_v1_conversion.go

```diff
@@ -0,0 +1,256 @@
+// Docker Manifest v2 Schema 1 Support
+// see also:
+// - https://docs.docker.com/registry/spec/manifest-v2-1/
+// - https://github.com/moby/moby/blob/master/image/v1/imagev1.go
+// - https://github.com/containerd/containerd/blob/main/remotes/docker/schema1/converter.go
+// *************************************************************************************
+
+const (
+	MediaTypeDockerV2Schema1Manifest       = "application/vnd.docker.distribution.manifest.v1+json"
+	MediaTypeDockerV2Schema1SignedManifest = images.MediaTypeDockerSchema1Manifest
+	MediaTypeImageLayerZstd                = "application/vnd.oci.image.layer.v1.tar+zstd"
+)
+
+type FSLayer struct {
+	BlobSum digest.Digest `json:"blobSum"`
+}
+
+type History struct {
+	V1Compatibility string `json:"v1Compatibility"`
+}
+
+type V1Manifest struct {
+	FSLayers []FSLayer `json:"fsLayers"`
+	History  []History `json:"history"`
+}
+
+type V1History struct {
+	Author          string    `json:"author,omitempty"`
+	Created         time.Time `json:"created"`
+	Comment         string    `json:"comment,omitempty"`
+	ThrowAway       *bool     `json:"throwaway,omitempty"`
+	Size            *int      `json:"Size,omitempty"`
+	ContainerConfig struct {
+		Cmd []string `json:"Cmd,omitempty"`
+	} `json:"container_config,omitempty"`
+}
+
+func ConvertV1ManifestToV2(ctx context.Context, c *client, ref string, v1ManifestDesc ocispecv1.Descriptor) (ocispecv1.Descriptor, error) {
+	buf := bytes.NewBuffer([]byte{})
+	if err := c.Fetch(ctx, ref, v1ManifestDesc, buf); err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to fetch v1 manifest blob: %w", err)
+	}
+
+	var v1Manifest V1Manifest
+	if err := json.Unmarshal(buf.Bytes(), &v1Manifest); err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to unmarshal v1 manifest: %w", err)
+	}
+
+	layers, diffIDs, history, err := ParseV1Manifest(ctx, c, ref, &v1Manifest)
+	if err != nil {
+		return ocispecv1.Descriptor{}, err
+	}
+
+	configDesc, configBytes, err := CreateConfig(&v1Manifest, diffIDs, history)
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to create config: %w", err)
+	}
+
+	v2ManifestDesc, v2ManifestBytes, err := CreateV2Manifest(configDesc, layers)
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to create v2 manifest: %w", err)
+	}
+
+	err = c.cache.Add(configDesc, io.NopCloser(bytes.NewReader(configBytes)))
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to write config blob to cache: %w", err)
+	}
+
+	err = c.cache.Add(v2ManifestDesc, io.NopCloser(bytes.NewReader(v2ManifestBytes)))
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to write manifest blob to cache: %w", err)
+	}
+
+	return v2ManifestDesc, nil
+}
+
+func ParseV1Manifest(ctx context.Context, c *client, ref string, v1Manifest *V1Manifest) (layers []ocispecv1.Descriptor, diffIDs []digest.Digest, history []ocispecv1.History, err error) {
```

Comment:
the function is exported but uses internal parameters.
I would suggest to keep it public but only use exported interfaces like `Client` or `Cache`.